Sources/SwiftFusion/Inference/PenguinExtensions.swift

```diff
@@ -104,6 +137,577 @@ extension ArrayBuffer: CustomDebugStringConvertible {
+          self_.storage[unsafelyAssuming: Type<C>()].distance(
+            from: start.storage[Type<C.Index>()], to: end.storage[Type<C.Index>()])
+        },
+        // Hidden collection requirements
+        customIndexOfEquatableElement: { self_, e in
+          self_.storage[unsafelyAssuming: Type<C>()]
+            ._customIndexOfEquatableElement(e).map { $0.map(Index.init) }
+        },
+        customLastIndexOfEquatableElement: { self_, e in
+          self_.storage[unsafelyAssuming: Type<C>()]._customLastIndexOfEquatableElement(e)
+            .map { $0.map(Index.init) }
+        },
+        // MutableCollection requirements.
+        swapAt: { self_, i, j in 
+          self_.storage[unsafelyAssuming: Type<C>()]
+            .swapAt(i.storage[Type<C.Index>()], j.storage[Type<C.Index>()])
+        }
+      )
+    }
+  }
+  
+  public var asAny: Any { storage.asAny }
+}
+
+extension AnyMutableCollection: Sequence {
+  public struct Iterator: IteratorProtocol {
+    var storage: Storage
+    // There's only one operation, so probably no point bothering with a vtable.
+    var next_: (inout Self) -> Element?
+
+    init<T: IteratorProtocol>(_ x: T) where T.Element == Element {
+      storage = .init(x)
+      next_ = { $0.storage[unsafelyAssuming: Type<T>()].next() }
+    }
+
+    public mutating func next() -> Element? { next_(&self) }
+  }
+
+  public func makeIterator() -> Iterator { vtable[0].makeIterator(self) }
+
+  /// If `self` supports a contiguous storage representation, assumes that representation and
+  /// returns the result of passing the storage to `body`; returns `nil` otherwise.
+  ///
+  /// - Note: this documentation might be wrong: https://bugs.swift.org/browse/SR-13486
+  public func withContiguousStorageIfAvailable<R>(
+    _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R? {
+    var result: R? = nil
+    try vtable[0].existentialDispatch.withContiguousStorageIfAvailable(storage) {
+      try result = body(
+        UnsafeBufferPointer(start: $0?.assumingMemoryBound(to: Element.self), count: $1))
+    }
+    return result
+  }
+
+  public func _customContainsEquatableElement(_ e: Element) -> Bool? {
+    vtable[0].customContainsEquatableElement(self, e)
+  }
+
+  public __consuming func _copyToContiguousArray() -> ContiguousArray<Element> {
+    vtable[0].copyToContiguousArray(self)
+  }
+  
+  public __consuming func _copyContents(
+    initializing target: UnsafeMutableBufferPointer<Element>
+  ) -> (Iterator,UnsafeMutableBufferPointer<Element>.Index) {
+    vtable[0].copyContents(self, target)
+  }
+}
+
+extension AnyMutableCollection: MutableCollection {
+  public struct Index: Comparable {
+    fileprivate var storage: Storage
+    // Note: using a vtable here actually made benchmarks slower.  We might want to try again if we
+    // expand what the benchmarks are doing.  My hunch is that the cost of copying an Index is less
+    // important than the cost of checking for equality.
```

Comment:
It's not immediately obvious to me why the vtable makes checking for equality slower. Could you explain why? Is it: The vtable makes the equality checking slower because it has to follow an extra pointer every time it makes the function call?

Reply:
Yes exactly.  If you have an idea of how to clarify that in the comment, please leave a suggestion.